pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,146 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
+    # is less than `slope_max` and the value of the daytime
+    # power curve is greater than `power_min` times the median of the
+    # daytime power curve.
+    #
+    # Based on the PVFleets QA Analysis project
+    if not freq:
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds / 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds / 60
+
+    # Use the slope of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(
+        ac_power, power_quantile, frequency_quantile
+    )
+    normalized_power = powercurve / powercurve.max()
+    power_slope = (normalized_power.diff()
+                   / normalized_power.index.to_series().diff()) * freq
+
+    clipped_times = _clipped(powercurve, power_slope,
+                             powercurve.median() * power_min,
+                             slope_max)
+    clipping_cumsum = (~clipped_times).cumsum()
+    # get the value of the cumulative sum of the longest True span
+    longest_clipped = clipping_cumsum.value_counts().idxmax()
+    # select the longest span that satisfies the clipping criteria
+    longest = powercurve[clipping_cumsum == longest_clipped]
+
+    if longest.index[-1] - longest.index[0] >= 60:
+        # if the period of clipping is at least 60 minutes then we
+        # have enough evidence to determine the clipping threshold.
+        return longest.mean()
+    return None
+
+
+def threshold(ac_power, slope_max=0.0035, power_min=0.75,
+              power_quantile=0.995, frequency_quantile=0.25,
+              freq=None):
+    """Detect clipping based on a maximum power threshold.
+
+    A clipping threshold is calculated from the AC power data and any
+    power greater than the threshold is flagged as clipped.
+
+    To calculate the clipping threshold, `ac_power` is aggregated at
+    each minute of the day. Low power data is removed to eliminate
+    night-time periods and the 99.5% quantile is computed at each
+    minute. If the absolute value of the slope of the 99.5%
+    quantile is less than `slope_max` for
+    a continuous period of at least one hour then clipping is
+    indicated. The mean power for that period is used as the
+    threshold. If there are multiple periods with a slope less
+    than `slope_max` then the longest period is used to compute
+    the threshold.
+
+    Parameters
+    ----------
+    ac_power : Series
+        DatetimeIndexed series of AC power data.
+    slope_max : float, default 0.0035
+        Maximum absolute value of slope of AC power quantile for
+        clipping to be indicated. The default value has been derived
+        empirically to prevent false positives for tracking PV
+        systems.
+    power_min: float, default 0.75
+        The power during periods with slope less than `slope_max` must
+        be greater than `power_min` time the median daytime power.
```

Comment:
The description of this parameter needs to be improved. I'm having a hard time finding the words at the moment, will come back to it later.